examples/fit-model-ceres.cpp

```diff
@@ -18,61 +18,240 @@
  * limitations under the License.
  */
 #define GLM_FORCE_UNRESTRICTED_GENTYPE
-#include "eos/core/Landmark.hpp"
-#include "eos/core/LandmarkMapper.hpp"
-#include "eos/core/read_pts_landmarks.hpp"
-#include "eos/fitting/ceres_nonlinear.hpp"
-#include "eos/fitting/contour_correspondence.hpp"
-#include "eos/fitting/fitting.hpp"
-#include "eos/morphablemodel/Blendshape.hpp"
 
-#include "Eigen/Core"
-
-#include "glm/ext.hpp"
-#include "glm/glm.hpp"
-
-#include "ceres/ceres.h"
-#include "ceres/cubic_interpolation.h"
-#include "ceres/rotation.h"
+#include <chrono>
+#include <iostream>
+#include <stdexcept>
+#include <vector>
 
-#include "opencv2/core/core.hpp"
+#include "boost/filesystem.hpp"
+#include "boost/program_options.hpp"
 #include "opencv2/imgproc/imgproc.hpp"
 #include "opencv2/highgui/highgui.hpp"
 
-#include "boost/filesystem.hpp"
-#include "boost/program_options.hpp"
+#include "eos/core/Landmark.hpp"
+#include "eos/core/LandmarkMapper.hpp"
+#include "eos/core/read_pts_landmarks.hpp"
+#include "eos/fitting/ceres_nonlinear.hpp"
+#include "eos/morphablemodel/Blendshape.hpp"
 
-#include <chrono>
-#include <iostream>
-#include <vector>
 
 using namespace eos;
-using namespace glm;
 using namespace ceres;
 namespace po = boost::program_options;
 namespace fs = boost::filesystem;
 using eos::core::Landmark;
 using eos::core::LandmarkCollection;
+using eos::core::IndexedLandmark;
+using eos::core::IndexedLandmarkCollection;
 using cv::Mat;
 using Eigen::Vector2f;
-using std::cout;
-using std::endl;
-using std::string;
-using std::vector;
-
-// print a vector:
-template <typename T>
-std::ostream& operator<<(std::ostream& out, const std::vector<T>& v)
-{
-    if (!v.empty())
-    {
+
+
+#ifndef EOS_CERES_EXAMPLE_USE_PERSPECTIVE
+#define EOS_CERES_EXAMPLE_USE_PERSPECTIVE true
+#endif
+
+#ifndef EOS_CERES_EXAMPLE_SHAPES_NUM
+#define EOS_CERES_EXAMPLE_SHAPES_NUM 63
+#endif
+
+#ifndef EOS_CERES_EXAMPLE_BLENDSHAPES_NUM
+#define EOS_CERES_EXAMPLE_BLENDSHAPES_NUM 6
+#endif
+
+#ifndef EOS_CERES_EXAMPLE_COLOR_COEFFS_NUM
+#define EOS_CERES_EXAMPLE_COLOR_COEFFS_NUM 10
+#endif
+
+
+namespace eos {
+    namespace ceres_example {
+        struct HelpCallException : public std::exception {
+            const char *what() const noexcept override {
+                return "User called help";
+            }
+        };
+
+        struct CliArguments {
```

Comment:
I think I'd like to keep the structure of the command-line argument handling like it is in the other examples now. It might in principle be a good idea to refactor those things into methods like you did, but I think particularly the introduction of a new type (`CliArguments`) and even an exception make the code much less trivial and harder to read for beginners. The example is really meant as a "simple example", and even if the `main` gets a bit long, I think it's easier to reason about than introducing new types/functions.

Reply:
It was made to decompose a too big main function. Do you think it is better to put all this code together? Maybe it is better to leave this function, but pass all that it returns as in-out arguments? What do you think? It will avoid new `struct` creation and will save code decomposition.